data_prep/project_context/context_introspector.py

```diff
@@ -0,0 +1,203 @@
+
+    if lang == 'c++':
+      return sig + ';'
+    if lang == 'c':
+      return 'extern "C" ' + sig + ';'
+
+    print('Unsupported declaration requested')
+    return ''
+
+  def _get_typedef_type(self, info: dict) -> str:
+    """Reconstructs type definition for a typedef element.
+    Simply queries for a single line in source code."""
+    file_name = os.path.normpath(info['source']['source_file'])
+    typedef_line = info['source']['source_line']
+    source_line = introspector.query_introspector_source_code(
+        self._benchmark.project, file_name, typedef_line, typedef_line)
+    # Check to ensure typedef is within the source line.
+    # There are instances where it isn't.
+    # One example, is when it is combined with a struct declaration
+    # 1. typedef X {
+    # 2.   T elem1;
+    # 3. } X;
+    # The typedef element has value 3. In this case,
+    # it would be better to ignore the string.
+    if 'typedef' not in source_line:
+      return ''
+
+    return source_line
+
+  def _get_struct_type(self, info: dict, seen_types: set,
+                       types_to_get: set) -> str:
+    """Reconstructs type definition from a struct element.
+    Also adds newly seen types to recursively collect definitions."""
+    # Attempt to recursively retrieve type information for elements
+    elements = info['elements']
+    file_name = os.path.normpath(info['source']['source_file'])
+    begin_line = int(info['source']['source_line'])
+    begin_line_elem = int(elements[0]['source']['source_line'])
+    end_line_elem = int(elements[-1]['source']['source_line'])
+    curr_line = int(begin_line)
+
+    reconstructed_type = ''
+
+    # The reason we need to iteratively query FI is because
+    # we have to parse each individual source line and
+    # extract a possible type which we will recursively
+    # query the definition for.
+    while curr_line < begin_line_elem:
+      source_line = introspector.query_introspector_source_code(
+          self._benchmark.project, file_name, curr_line, curr_line)
+      curr_line += 1
+      reconstructed_type += source_line
+
+    while curr_line <= end_line_elem:
+      source_line = introspector.query_introspector_source_code(
+          self._benchmark.project, file_name, curr_line, curr_line)
+
+      reconstructed_type += source_line
+
+      newly_seen_type = self._clean_type(
+          self._extract_type_from_source_line(
+              source_line, elements[curr_line - begin_line_elem]))
+
+      # If we do not see a '}' in the final element's source line
+      # Then we can add it ourselves. This would cause problems
+      # when typedef and struct definitions are combined.
+      # The alternative is to query for source code lines until a '}' is found.
+      if curr_line == end_line_elem:
+        if '}' not in source_line:
+          reconstructed_type += '};\n'
+
+      curr_line += 1
+
+      if not newly_seen_type or newly_seen_type in seen_types:
```

Comment:
sorry, newly_seen_type is badly named. it should be `potential_new_type`, as it could be an old type. in the case where it was a previously seen type, it could be added to types_to_get and be looked up again redundantly.